### PR TITLE
Extract Migrator#create and #drop sql into separate methods

### LINKED ISSUE #319: Add `Granite::Migrator#create_sql` and `#drop_sql`
### tl;dr

`Granite::Migrator` generates SQL on `#drop` and `#create`, and runs it imminently. The possibility of having the migrator export the generated SQL would be nice for use in migration tools like `micrate`. The generating functionality is already in place, implementing this would just be moving the generating code into `create`- and `drop`- `_sql` methods (names debatable), and using them in place of the current generating code.

---

I'm using Granite for building a program (i.e. no amber), I'm also quite new to SQL, which is one of the reasons I enjoy Granite so much. 

Following the Granite [docs on migrating][docs], I stumble upon SQL in the micrate up and down sections, I do not know how Granite interfaces with its tables. Therefore I do not know how to define these migrations, and I take a look at the alternative `Granite::Migrator#create` and `#drop`, which both work great, however they seem wholly impractical to use long term, at least compared to `micrate`. 

I hoped I could extract the SQL generated by `Granite::Migrator`, but seems that's not possible, hence I ask here whether it's a good idea or not to extract the generated code into separate methods. Practical or not it adheres better to the single responsibility principle... Sorry for the spaghetti text.

I'll happily implement this if it's accepted as a good idea.

[docs]: https://github.com/amberframework/granite/blob/master/docs/migrations.md#creating-a-migration

### PR DESCRIPTION
**Pull request by Jens0512**
Fixes #319. 

I haven't managed to run the specs (difficulties with docker), but the changes shouldn't change any behaviour, only expose functionality.

### CODE REVIEW

**Comment by Jens0512** [comment]
But alas, as per Murphy's Law, "If something can go wrong, it will." I did. Fix coming right up

**Comment by Jens0512** [comment]
There, obvious mistake on my end :sweat_smile:. Fixed now

**Comment by robacarp** [comment]
@Jens0512 This is a nice added feature. What do you think about adding a test to show that these methods are working as expected?

**Comment by Jens0512** [comment]
Well, while that's a great idea, I'm not quite great enough at SQL to know what the should return...

**Comment by Blacksmoke16** [comment]
@Jens0512 Its pretty easy.  It would however depend on what the current adapter is.  Assuming for a model like:

```crystal
class Post < Granite::Base
  adapter pg
  table_name posts

  field name : String
  field body : String
  timestamps
end
```
* MySQL
  * Drop - ``DROP TABLE IF EXISTS `posts`;``
  * Create - ``CREATE TABLE `posts`( `id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY, `name` VARCHAR(255), `body` VARCHAR(255), `created_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP, `updated_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP)`` 
* PG
  * Drop - `DROP TABLE IF EXISTS "posts";`
  * Create - ``CREATE TABLE "posts"( "id" BIGSERIAL, "name" VARCHAR(255), "body" VARCHAR(255), "created_at" TIMESTAMP, "updated_at" TIMESTAMP)`` 
* SqLite
  * Drop - `DROP TABLE IF EXISTS "posts";`
  * Create - ``CREATE TABLE "posts"( "id" INTEGER NOT NULL, "name" VARCHAR(255), "body" VARCHAR(255), "created_at" VARCHAR, "updated_at" VARCHAR)``

**Comment by robacarp** [comment]
@Blacksmoke16 is right about the technique "run the method, copy the output, and assert that the result is always the output".

Typically, I wouldn't say that's a good practice, but in this case we already know it's working (if only by experience). So asserting that the output is consistent will do the job of ensuring that no accidental changes to the output of the methods are introduced without raising a warning.

**Comment by Jens0512** [comment]
> "run the method, copy the output, and assert that the result is always the output"

Sounds good to me, as long as we know the initial output is correct.

Thanks a ton for the help @Blacksmoke16. 

> Ideally we would use a model that has all data types in it to fully test the migrator.

Shall I create one and put it in `spec/spec_models.cr` then?

**Comment by Blacksmoke16** [comment]
@Jens0512 Yea that would be fine.  Be sure to also test the PG array types.  There is already a model that has them defined.  Will have to scope it to run on PG only since the other DBs dont support it.  Can reference https://github.com/amberframework/granite/blob/master/spec/granite_spec.cr#L261-L321

**Comment by Jens0512** [comment]
(Seems there was a model with all types already: [Review](https://github.com/amberframework/granite/blob/master/spec/spec_models.cr#L133))

**Comment by Jens0512** [comment]
The formatting of the generated SQL is terrible to behold haha

**Comment by Jens0512** [comment]
Oh no, ameba has detected a terrible line of whitespace...

**Comment by Jens0512** [comment]
Maybe mention somewhere in the README that you should run ameba before you push. Just in case you put a 2-3 extra white spaces anywhere... :D

**Comment by Blacksmoke16** [comment]
@Jens0512 This is looking great thanks!  I remembered there would be two other cases we should test.  

1. Natural PK, can use https://github.com/amberframework/granite/blob/master/spec/spec_models.cr#L190
2. UUID PK, can use https://github.com/amberframework/granite/blob/master/spec/spec_models.cr#L351

Would you mind adding specs for those as well?

**Comment by Blacksmoke16** [comment]
Sorry Satisfactory sidetracked me this weekend.  I'll get this reviewed tonight.

@robacarp @eliasjpr If you guys have any input as well, it would be nice.

**Comment by Jens0512** [comment]
Nothing to be sorry about :)

File: spec/granite/migrator/migrator_spec.cr [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+          "id" BIGSERIAL PRIMARY KEY,
+          "name" VARCHAR(255)
+          ,
+          "downvotes" INT
+          ,
+          "upvotes" BIGINT
+          ,
+          "sentiment" FLOAT
+          ,
+          "interest" REAL
```
Comment (Blacksmoke16) [inline spec/granite/migrator/migrator_spec.cr]: Im thinking the type of `interest` should be `DOUBLE PRECISION`.  Based on the [docs](https://www.postgresql.org/docs/9.5/datatype.html#DATATYPE-TABLE) that would be their 8 byte floating point type.

File: spec/granite/migrator/migrator_spec.cr [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        Review.migrator.create_sql.should eq <<-SQL
+          CREATE TABLE "reviews"(
+          "id" BIGSERIAL PRIMARY KEY,
+          "name" VARCHAR(255)
+          ,
+          "downvotes" INT
+          ,
+          "upvotes" BIGINT
+          ,
+          "sentiment" FLOAT
```
Comment (Blacksmoke16) [inline spec/granite/migrator/migrator_spec.cr]: Im thinking the type of `sentiment` should be `REAL`.  Based on the [docs](https://www.postgresql.org/docs/9.5/datatype.html#DATATYPE-TABLE) that would be their 4 byte floating point type.

File: spec/granite/migrator/migrator_spec.cr [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+          ,
+          "interest" REAL
+          ,
+          "published" BOOL
+          ,
+          "created_at" TIMESTAMP
```
Comment (Blacksmoke16) [inline spec/granite/migrator/migrator_spec.cr]: We _could_ use `TIMESTAMPZ` which would do conversion of the datetime object and store it in UTC.  However that would probably be a breaking change as people's datetime's would start returning differently then.  Can save that for another discussion, related to #225 .

File: spec/granite/migrator/migrator_spec.cr [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    end
+  end
+
+  describe "#create_sql" do
+    it "generates correct SQL with #{{{ env("CURRENT_ADAPTER") }}} adapter" do
+      {% if env("CURRENT_ADAPTER") == "pg" %}
+        Review.migrator.create_sql.should eq <<-SQL
+          CREATE TABLE "reviews"(
+          "id" BIGSERIAL PRIMARY KEY,
+          "name" VARCHAR(255)
```
Comment (Blacksmoke16) [inline spec/granite/migrator/migrator_spec.cr]: Should probably make `name` be `TEXT`, seems to be the preferred variable length string type for PG.

File: spec/granite/migrator/migrator_spec.cr [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+          `id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
+          `name` VARCHAR(255)
+          ,
+          `downvotes` INT
+          ,
+          `upvotes` BIGINT
+          ,
+          `sentiment` FLOAT
+          ,
+          `interest` REAL
```
Comment (Blacksmoke16) [inline spec/granite/migrator/migrator_spec.cr]: Can make this `DOUBLE` based on [this chart](http://www.peachpit.com/articles/article.aspx?p=1752305&seqNum=2)

File: spec/granite/migrator/migrator_spec.cr (hunk 3 above)
Comment (Jens0512) [reply spec/granite/migrator/migrator_spec.cr]: So I'll leave this untouched?

Comment (Blacksmoke16) [reply spec/granite/migrator/migrator_spec.cr]: Yea.

**Comment by Jens0512** [comment]
Thanks for the feedback, anything else?

**Pull request merged**